mesonbuild/build.py

```diff
@@ -2328,6 +2328,8 @@ def process_kwargs(self, kwargs):
             else:
                 raise InvalidArguments(f'Invalid rust_crate_type "{rust_crate_type}": must be a string.')
             if rust_crate_type == 'proc-macro':
+                if not kwargs.get('native', False):
+                    raise InvalidArguments('Rust "proc-macro" crate type requires "native: true".')
```

Comment:
So it looks like this would prospectively error out on projects that have never attempted to support cross building, and are currently being compiled natively, because they are doing something technically wrong despite that it still compiles okay on native-only builds. Am I understanding correctly?

Reply:
Yes. Would you prefer a warning instead?

Reply:
I guess that would be better, for backwards compat. Even if the current way is just wrong :)

Reply:
I think that would be better, yes.

Keeping in mind that these projects cannot be cross compiled at *all* on older versions of Meson, so people were unlikely to bother setting that either way...

Whatever else, we want that existing builds which worked, should continue to work due to stability of the semver major release version.

Reply:
Changed accordingly

Reply:
I think we could do something like:
```python
if kwargs.get('native', False):
    if self.environment.is_cross_build():
        error(...)
    else:
        warning(...)
```
since compilation would have already failed in the cross build case anyway, but would succeed otherwise?

Reply:
Sounds correct, yes. That would also work for me, your call :)